csharp/src/Google.Protobuf.Test/RefStructCompatibilityTest.cs

```diff
@@ -0,0 +1,115 @@
+        /// <summary>
+        /// Checks that the generated code can be compiler with an old C# compiler.
+        /// The reason why this test is needed is that even though dotnet SDK projects allow to set LangVersion,
+        /// this setting doesn't accurately simulate compilation with an actual old pre-roslyn C# compiler.
+        /// For instance, the "ref struct" types are only supported by C# 7.2 and higher, but even if
+        /// LangVersion is set low, the roslyn compiler still understands the concept of ref struct
+        /// and silently accepts them. Therefore we try to build the generated code with an actual old C# compiler
+        /// to be able to catch these sort of compatibility problems.
+        /// </summary>
+        [Test]
+        public void GeneratedCodeCompilesWithOldCsharpCompiler()
+        {
+            if (Environment.OSVersion.Platform != PlatformID.Win32NT)
+            {
+                // This tests needs old C# compiler which is only available on Windows. Skipping it on all other platforms.
+                return;
+            }
+
+            var currentAssemblyDir = Path.GetDirectoryName(typeof(RefStructCompatibilityTest).GetTypeInfo().Assembly.Location);
+            var testProtosProjectDir = Path.GetFullPath(Path.Combine(currentAssemblyDir, "..", "..", "..", "..", "Google.Protobuf.Test.TestProtos"));
+            var testProtosOutputDir = (currentAssemblyDir.Contains("bin/Debug/") || currentAssemblyDir.Contains("bin\\Debug\\")) ? "bin\\Debug\\net45" : "bin\\Release\\net45";
+
+            // If "ref struct" types are used in the generated code, compilation with an old compiler will fail with the following error:
+            // "XYZ is obsolete: 'Types with embedded references are not supported in this version of your compiler.'"
+            // We build the code with GOOGLE_PROTOBUF_REFSTRUCT_COMPATIBILITY_MODE to avoid the use of ref struct in the generated code.
+            var compatibilityFlag = "-define:GOOGLE_PROTOBUF_REFSTRUCT_COMPATIBILITY_MODE";
+            var sources = "*.cs";  // the generated sources from the TestProtos project
+            var args = $"-langversion:3 -target:library {compatibilityFlag} -reference:{testProtosOutputDir}\\Google.Protobuf.dll -out:{testProtosOutputDir}\\TestProtos.RefStructCompatibilityTest.OldCompiler.dll {sources}";
+            RunOldCsharpCompilerAndCheckSuccess(args, testProtosProjectDir);
+        }
+
+        /// <summary>
+        /// Invoke an old C# compiler in a subprocess and check it finished successful.
+        /// </summary>
+        /// <param name="args"></param>
+        /// <param name="workingDirectory"></param>
+        private void RunOldCsharpCompilerAndCheckSuccess(string args, string workingDirectory)
+        {  
+            using (var process = new Process())
+            {
+                process.StartInfo.FileName = "c:\\Windows\\Microsoft.NET\\Framework\\v4.0.30319\\csc.exe";  // Old C# 5 compiler from .NET framework
```

Comment:
Hard coded path is very brittle. Windows could be installed to a different path or drive.

Should look this up from an environmental variable.

Reply:
https://stackoverflow.com/questions/6660512/how-to-get-csc-exe-path

I think this would work if the test only runs with the net45 target:
```cs
using System.Runtime.InteropServices;
var frameworkPath = RuntimeEnvironment.GetRuntimeDirectory();
var cscPath = Path.Combine(frameworkPath, "csc.exe");
```

Alternatively the %WINDIR% environment variable would be an improvement. To be more exact and lookup the exact path would probably require accessing Windows registry.

Reply:
Good point. I updated the test to use the WINDIR env. It's not 100% but seems good enough for our purposes.